### ISSUE TITLE
4 SPACES FOR ANYONE!

### ISSUE THREAD

**Issue by MishUshakov** [vuejs/vetur#579]
- [x] I have searched through existing issues
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: macOs
- Vetur version: latest
- VS Code version: latest

## Problem

Impossible to write code, when using 4 spaces, please fix it. Very unsatisfied!

## Reproducible Case

Use 4 Spaces in your editor instead of 2 and see the red blood flowing on your code

<img width="1680" alt="screen shot 2017-12-14 at 21 42 52" src="https://user-images.githubusercontent.com/10400064/34013491-ff7ed984-e117-11e7-8090-328685c8341b.png">

**Comment by cotej** [comment]
Yep, same here - looks like the latest update implicitly forced some stricter linting down own throats...

**Comment by TongZhangzt** [comment]
Same problem here. Got the problem after upgrade to 0.11. 4.

**Comment by octref** [comment]
I don't see it happening in https://github.com/octref/veturpack. Do you have a repo where I can try reproing this issue? (Or try modifying veturpack to repro it).

![image](https://user-images.githubusercontent.com/4033249/34014403-db3abe6e-e0cf-11e7-9072-4196bc97a886.png)

**Comment by cotej** [comment]
@octref Likely that's because you have explicit ESLint configuration in that repo and have set `"vetur.validation.template": false` in your vscode settings.  The issue occurs when using vetur's template validation.

**Comment by cotej** [comment]
FYI for the others affected by this, here's what you can do to rollback.

Download the previous version of the Vetur extension from here:
https://octref.gallery.vsassets.io/_apis/public/gallery/publisher/octref/extension/vetur/0.11.3/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage?install=true

Then in VSCode, on the Extension pane, click the '...' button
  -> Disable Auto Updating Extensions (so we don't wind up back at 0.11.4 automatically)
  -> Install from VSIX... (selected the file you just download)

Then just wait for the 'Reload Window' notification... done!

**Comment by MishUshakov** [comment]
@cotej and @octref @TongZhangzt  Thank you guys for the support! Hope next update will be pushed soon, so we can continue coding well again. Also noticed very strict rules on v-for tags, it requires me to set unique key, but I don't need or doesn't have to set them because it is not an actual error, so if vue team is reading this, please fix it, since it is just a good recommendation, not a requirement.

**Comment by octref** [comment]
The rule is: https://github.com/vuejs/eslint-plugin-vue/blob/master/docs/rules/html-indent.md

If you do not want Vetur to lint your Vue template, turn it off with VS Code setting `vetur.validation.template: false`.

If you do want Vetur to lint your Vue template, but without this rule, take a look at Veturpack's config: https://github.com/octref/veturpack/blob/master/.eslintrc

You can add an entry to relax any rule you want. For this one, it could be:

```
"vue/html-indent": "off"
```

Basically, Vetur uses the default config for `eslint-plugin-vue` when you don't have one installed in your project. If you want to customize it, do this:

- `yarn add -D eslint eslint-plugin-vue eslint-config-vue`
- Setup an eslintrc like this: https://github.com/octref/veturpack/blob/master/.eslintrc

I'll add more docs later.

**Comment by GufNZ** [comment]
`npm WARN eslint-config-vue@2.0.2 requires a peer of eslint@^2.0.0 || ^3.0.0 but none is installed. You must install peer dependencies yourself.`
I have `"eslint": "^4.12.1"` installed.
Will that cause any issues?

**Comment by octref** [comment]
Umm, ok https://github.com/vuejs/eslint-config-vue is actually deprecated. I had no idea...

I'll look into this later today. For now you can choose either use 0.11.3 by installing from VSIX, or disable the template linting.

**Comment by GufNZ** [comment]
That might explain all the `Definition fir rule 'vue/something' was not found` errors I'm suddenly seeing.

**Comment by GufNZ** [comment]
I am noticing that all the definition not found errors are agains JS files, and not vue files.  Maybe I did something wrong when I configured ESLint?

**Comment by HerringtonDarkholme** [comment]
@octref I think we should simply change our default rule set to `essential` which only reports possible runtime error. So styling issue reporting like indentation and casing is opt-in. (by eslint-vscode plugin).

**Comment by octref** [comment]
This in an error from [`html-indent`](https://github.com/vuejs/eslint-plugin-vue/blob/master/docs/rules/html-indent.md) rule of `eslint-plugin-vue`.

In 0.11.4, I used the [`recommended`](https://github.com/vuejs/eslint-plugin-vue#priority-c-recommended-minimizing-arbitrary-choices-and-cognitive-overhead) set from `eslint-plugin-vue`, which includes `html-indent`. I've changed that to [`essential`](https://github.com/vuejs/eslint-plugin-vue#priority-a-essential-error-prevention) which does not include it.

So with 0.11.5, when you have `vetur.template.validate: true`, this won't be marked as error. I also updated the doc for configuring ESLint to use custom rules: https://vuejs.github.io/vetur/linting-error.html

I also upgrade https://github.com/octref/veturpack to use the newest `eslint-plugin-vue`. 

@GufNZ In the past, rule sets of `eslint-plugin-vue` lived at https://github.com/vuejs/eslint-config-vue. However now the rule live in `eslint-plugin-vue` itself, so you can do:

- `yarn add -D eslint eslint-plugin-vue@next`
- Write a `.eslintrc` like so

  ```json
  {
    "extends": [
      "eslint:recommended",
      "plugin:vue/recommended"
    ],
    "rules": {
      "vue/html-self-closing": "off"
    }
  }
  ```

Finally, updated doc for linting is at https://vuejs.github.io/vetur/linting-error.html.
Sorry for the problem.

**Comment by GufNZ** [comment]
So, I'm still a bit confused.  I have vetur 0.11.5; have removed the `vetur.validation.template: false` line from my settings; have setup `.eslintrc.js` as above, but the rules don't seem to be getting applied at all.

Not sure what to try next...

	module.exports = {
		root: true,

		parserOptions: {
			parser: 'babel-eslint',
			sourceType: 'module'
		},
		env: {
			browser: true
		},
		extends: ['airbnb-base', 'plugin:vue/recommended'],

		// Required to lint *.vue files:
		plugins: [
			'html',
			'vue'
		],

		// Check if imports actually resolve:
		settings: {
			'import/resolver': {
				webpack: {
					config: 'build/webpack.base.conf.js'
				}
			}
		},

		// Add your custom rules here:
		rules: {
			'linebreak-style': ['error', 'unix'],

			// Don't require .vue extension when importing:
			'import/extensions': [
				'error',
				'always',
				{
					js: 'never',
					vue: 'never'
				}
			],

	//......8<...........

			'vue/mustache-interpolation-spacing': ['error', 'never'],
			'vue/require-default-prop': 'error',
			'vue/require-prop-types': 'error',
			'vue/v-bind-style': ['warn', 'shorthand'],
			'vue/v-on-style': ['warn', 'shorthand'],
			'vue/html-quotes': ['error', 'double'],
			'vue/no-confusing-v-for-v-if': 'error',

	//......8<...........

		}
	}

**Comment by HerringtonDarkholme** [comment]
@GufNZ you also need to install vscode-eslint plugin. https://vuejs.github.io/vetur/linting-error.html

**Comment by GufNZ** [comment]
got vscode-eslint 1.4.3 already.

I was getting spotty linting - sometimes it would apply some rules, but normally none of the vue/* ones.

I will try again tomorrow with the recommended config from that link.

**Comment by GufNZ** [comment]
My user settings:

	"vetur.validation.template": false,
	"eslint.options": {
		"extensions": [".js", ".vue", ".html"]
	},
	"eslint.validate": [
		"javascript",
		"javascriptreact",
		{
			"language": "vue",
			"autoFix": true
		}
	],

**Comment by GufNZ** [comment]
@octref How was your holiday?  Meanwhile, any progress on this?

**Comment by octref** [comment]
@GufNZ Try removing the eslint-plugin-html and just use eslint-plugin-vue.

**Comment by GufNZ** [comment]
That plus restarting vsCode seems to have done it, at least for the one file I've checked so far.
